### PR TITLE
Horizontal executions merging fix

### LINKED ISSUE #559: OnTheFlyMerging duplicates LocalScalar definitions
The OnTheFlyMerging pass folds up HorizontalExecutions on the OIR. There are cases when it inlines a horizontal execution that already has LocalScalars in the `declarations` list for the `HorizontalExecution`, and when that's the case then the resulting top-level symbol table sees multiple definitions of the same `LocalScalar`.

The resolution to this to ensure that every `LocalScalar.name` is unique to every `HorizontalExecution`. To make this happen, the `OnTheFlyMerging` pass needs to add a step that looks for existing `LocalScalars` in the `HorizontalExecution` that it is merging in and map these to new names, with new `LocalScalars`.

### PR DESCRIPTION
**Pull request by jdahm**
## Description

Fix #559.

### CODE REVIEW

File: tests/test_unittest/test_gtc/test_passes/test_oir_optimizations/test_horizontal_execution_merging.py [hunk 1]
```diff
@@ -160,14 +160,20 @@ def test_on_the_fly_merging_repeated():
...
+                body=[AssignStmtFactory(left__name="out", right__name="tmp_offset")]
             ),
-            HorizontalExecutionFactory(body=[AssignStmtFactory(left__name="tmp")]),
             HorizontalExecutionFactory(
-                body=[AssignStmtFactory(left__name="out2", right__name="tmp")]
+                body=[AssignStmtFactory(left__name="out2", right__name="tmp_offset")]
             ),
         ],
-        declarations=[TemporaryFactory(name="tmp")],
+        declarations=[TemporaryFactory(name="tmp"), TemporaryFactory(name="tmp_offset")],
```
Comment (gronerl) [inline tests/test_unittest/test_gtc/test_passes/test_oir_optimizations/test_horizontal_execution_merging.py:176]: I propose to make this a new separate test, the old one relied on the particular read-write pattern that was there and I'm not sure if the new test would capture the old test properly.

File: src/gtc/passes/oir_optimizations/horizontal_execution_merging.py [hunk 2]
```diff
@@ -23,6 +23,44 @@
...
+
+    def visit_HorizontalExecution(
+        self, node: oir.HorizontalExecution, *, ctx: "Context"
+    ) -> oir.HorizontalExecution:
+        new_symbol_name = symbol_name_creator(ctx.used_names)
+        name_map = {
+            d.name: new_symbol_name(d.name) for d in node.declarations if d.name in ctx.used_names
+        }
+
+        ctx.used_names |= {d.name for d in node.declarations}
```
Comment (gronerl) [inline src/gtc/passes/oir_optimizations/horizontal_execution_merging.py:45]: shouldn't this add the new names instead of the old ones?

**Review by gronerl — CHANGES_REQUESTED** [review]

File: tests/test_unittest/test_gtc/test_passes/test_oir_optimizations/test_horizontal_execution_merging.py (hunk 1 above)
Comment (jdahm) [reply tests/test_unittest/test_gtc/test_passes/test_oir_optimizations/test_horizontal_execution_merging.py]: Sounds good. The problem was that this test didn't even claim to do what it said it was doing. It does "multiple" applications of the pass, but there is not a nested `_merge` call. I added this test back and renamed both.

File: src/gtc/passes/oir_optimizations/horizontal_execution_merging.py (hunk 2 above)
Comment (jdahm) [reply src/gtc/passes/oir_optimizations/horizontal_execution_merging.py]: Good catch! It should have a combination of the new (if collision in used_names) and old names (if no collision detected). I fixed this.

**Comment by jdahm** [comment]
@gronerl Take a look at #564. That might be the way to go instead of this PR. Thanks!

**Comment by jdahm** [comment]
I believe I now prefer #562 over this. Closing in favor of that one.

**Pull request closed without merging**